File: python/array_search/parametric_search_production.py
from collections.abc import Sequence


def _can_ship(weights: Sequence[int], days: int, capacity: int) -> bool:
    used_days = 1
    current_load = 0

    for weight in weights:
        if current_load + weight > capacity:
            used_days += 1
            current_load = 0
            if used_days > days:
                return False
        current_load += weight

    return True


def minimum_ship_capacity(weights: Sequence[int], days: int) -> int:
    if not weights:
        raise ValueError("weights must not be empty")
    if days <= 0:
        raise ValueError("days must be positive")

    low = max(weights)
    high = sum(weights)

    while low < high:
        mid = low + (high - low) // 2

        if _can_ship(weights, days, mid):
            high = mid
        else:
            low = mid + 1

    return low

File: java/array_search/ParametricSearchProduction.java
public class ParametricSearchProduction {
    private static boolean canShip(int[] weights, int days, int capacity) {
        int usedDays = 1;
        int currentLoad = 0;

        for (int weight : weights) {
            if (currentLoad + weight > capacity) {
                usedDays++;
                currentLoad = 0;
                if (usedDays > days) {
                    return false;
                }
            }
            currentLoad += weight;
        }

        return true;
    }

    public static int minimumShipCapacity(int[] weights, int days) {
        if (weights == null || weights.length == 0) {
            throw new IllegalArgumentException("weights must not be empty");
        }
        if (days <= 0) {
            throw new IllegalArgumentException("days must be positive");
        }

        int low = weights[0];
        int high = 0;

        for (int weight : weights) {
            low = Math.max(low, weight);
            high += weight;
        }

        while (low < high) {
            int mid = low + (high - low) / 2;

            if (canShip(weights, days, mid)) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }

        return low;
    }
}
